// raspi-pca9685-pwm/dist/pca9685pwm.js
"use strict";
/*
The MIT License (MIT)

Copyright (c) Kiyo Chinzei (kchinzei@gmail.com)

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.
*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.module = exports.PCA9685PWM = void 0;
/*
   Raspi-PCA9685-pwm - Hardware PWM by PCA9685 via I2C on the Raspberry Pi.

   Make Asayake to Wake Project.
   Kiyo Chinzei
   https://github.com/kchinzei/raspi-pca9685-pwm

   Raspi-pca9685-pwm is built upon raspi-i2c, submodule of Raspi.js
      https://www.npmjs.com/package/raspi
   I also looked at pca9685 module in npm.
      https://www.npmjs.com/package/pca9685
      https://github.com/101100/pca9685
   I also learned from Adafruit PCA9685 python code
      https://github.com/adafruit/Adafruit_Python_PCA9685/
   But the API is modified to be consistent to raspi-soft-pwm.
      https://github.com/nebrius/raspi-soft-pwm

   Due to this design, you can switch easily between
   raspi-soft-pwm and raspi-pca9685-pwm.
*/
const pca9685module_1 = require("./pca9685module");
const raspi_soft_pwm_1 = require("raspi-soft-pwm");
function checkPin(config) {
    // It preserves channel's current PWM status.
    // If application should init PWM before use, it's your task.
    let pin = 0;
    if (typeof config === 'object') {
        pin = config.pin;
    }
    else {
        pin = config;
    }
    return pin;
}
class PCA9685PWM {
    constructor(config) {
        this._ch = 0;
        this._board = 0;
        const pin = checkPin(config);
        let port = 0;
        if (typeof pin === 'number') {
            port = pin;
        }
        else {
            port = Number(pin);
            if (isNaN(port)) {
                throw new RangeError(`Invalid port number '${pin}', not a number string).`);
            }
        }
        if (port < 0 || port >= pca9685module_1.publicConst.maxChannelsPerBoard * pca9685module_1.publicConst.maxBoards) {
            throw new RangeError(`Invalid port number ${port}, out of [0,${pca9685module_1.publicConst.maxChannelsPerBoard * pca9685module_1.publicConst.maxBoards}).`);
        }
        this._ch = port % pca9685module_1.publicConst.maxChannelsPerBoard;
        this._board = Math.floor(port / pca9685module_1.publicConst.maxChannelsPerBoard);
        let frequency = pca9685module_1.publicConst.defaultFrequency;
        if (typeof config === 'object' && typeof config.frequency === 'number') {
            frequency = config.frequency;
        }
        if (typeof PCA9685PWM._pca9685[this.board] === 'undefined') {
            PCA9685PWM._pca9685[this.board] = new pca9685module_1.PCA9685Module(this.board, frequency);
        }
        this.read();
    }
    get ch() { return this._ch; }
    get board() { return this._board; }
    get frequency() { return PCA9685PWM._pca9685[this.board].frequency; }
    get dutyCycle() { return this.read(); }
    /* For SoftPWM compatibility */
    /* istanbul ignore next */
    get range() { return pca9685module_1.publicConst.stepsPerCycle; }
    /* istanbul ignore next */
    get pins() { return this.ch + this.board * pca9685module_1.publicConst.maxChannelsPerBoard; }
    write(dutyCycle) {
        PCA9685PWM._pca9685[this.board].setDutyCycle(this.ch, dutyCycle);
    }
    read() {
        return PCA9685PWM._pca9685[this.board].dutyCycle(this.ch);
    }
    on() {
        PCA9685PWM._pca9685[this.board].channelOn(this.ch);
    }
    off() {
        PCA9685PWM._pca9685[this.board].channelOff(this.ch);
    }
    allOff() {
        PCA9685PWM._pca9685[this.board].channelOff();
    }
    /* istanbul ignore next */
    destroy() {
        // It does not destroy I2C communication.
    }
}
exports.PCA9685PWM = PCA9685PWM;
PCA9685PWM._pca9685 = new Array(pca9685module_1.publicConst.maxBoards);
exports.module = {
    /*
      Factory function that return either SoftPWM or PCA9635PWM.
      if given pin is a non-number string (e.g., 'GPIO22'), SoftPWN.
      if gicen pin represents a number (e.g., 22 or '022'), PCA9685PWM.
    */
    createPWM(config) {
        const pin = checkPin(config);
        if (typeof pin === 'string' && isNaN(Number(pin))) {
            return new raspi_soft_pwm_1.SoftPWM(config);
        }
        return new PCA9685PWM(config);
    }
};
